pkg/signer/memory.go

```diff
@@ -39,33 +41,53 @@ type Memory struct {
 	signature.ECDSASignerVerifier
 }
 
-// create a self-signed CA and generate a timestamping certificate to rekor
-func NewTimestampingCertWithSelfSignedCA(pub crypto.PublicKey) ([]*x509.Certificate, error) {
-	// generate self-signed CA
-	caPrivKey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+// Generate a timestamping certificate for pub using the signer. The chain must verify the signer's public key if provided.
+// Otherwise, a self-signed root CA will be generated.
+func NewTimestampingCertWithChain(ctx context.Context, pub crypto.PublicKey, signer signature.Signer, chain []*x509.Certificate) ([]*x509.Certificate, error) {
+	// Get the signer's (rekor's) public key
+	signerPubKey, err := signer.PublicKey(options.WithContext(ctx))
 	if err != nil {
-		return nil, errors.Wrap(err, "generating private key")
+		return nil, err
 	}
-	ca := &x509.Certificate{
-		SerialNumber: big.NewInt(2019),
-		Subject: pkix.Name{
-			Organization:  []string{"Root CA Test"},
-			Country:       []string{"US"},
-			Province:      []string{""},
-			Locality:      []string{"San Francisco"},
-			StreetAddress: []string{"Golden Gate Bridge"},
-			PostalCode:    []string{"94016"},
-		},
-		NotBefore:             time.Now(),
-		NotAfter:              time.Now().AddDate(10, 0, 0),
-		IsCA:                  true,
-		KeyUsage:              x509.KeyUsageDigitalSignature | x509.KeyUsageContentCommitment | x509.KeyUsageCertSign,
-		BasicConstraintsValid: true,
+
+	// If the signer is not in-memory, retrieve the crypto.Signer
+	var cryptoSigner crypto.Signer
+	if s, ok := signer.(*gcp.SignerVerifier); ok {
+		if cryptoSigner, _, err = s.CryptoSigner(ctx, func(err error) {}); err != nil {
+			return nil, errors.Wrap(err, "getting kms signer")
+		}
+	} else {
+		cryptoSigner = signer.(crypto.Signer)
 	}
-	caBytes, err := x509.CreateCertificate(rand.Reader, ca, ca, &caPrivKey.PublicKey, caPrivKey)
-	if err != nil {
-		return nil, err
+
+	if len(chain) == 0 {
+		// Generate a self-signed root CA.
+		ca := &x509.Certificate{
+			SerialNumber: big.NewInt(2019),
+			Subject: pkix.Name{
+				Organization:  []string{"in-memory root CA"},
+				Country:       []string{"US"},
+				Province:      []string{""},
+				Locality:      []string{"San Francisco"},
```

Comment:
Should we fill this in with slightly more real info?

Reply:
I just removed the fake US thing and called it a rekor in-memory CA